src/NodeApi/Interop/JSSynchronizationContext.cs

```diff
@@ -42,6 +43,148 @@ public virtual void Dispose()
     public abstract void OpenAsyncScope();
 
     public abstract void CloseAsyncScope();
+
+    public void Post(Action action, bool allowSync = false)
+    {
+        if (allowSync && Current == this)
+        {
+            action();
+        }
+        else
+        {
+            Post(_ =>
+            {
+                if (IsDisposed) return;
+                action();
+            }, null);
+        }
+    }
+
+    public void Post(Func<Task> asyncAction, bool allowSync = false)
+    {
+        if (allowSync && Current == this)
+        {
+            _ = asyncAction();
+        }
+        else
+        {
+            Post((_) =>
+            {
+                if (IsDisposed) return;
+                _ = asyncAction();
+            }, null);
+        }
+    }
+
+    public void Run(Action action)
```

Comment:
>Run [](http://example.com/codeflow?start=16&length=3)

I am trying to understand the need for the new set of Run* functions. Why do we need them in the base JSSynchronizationContext?

Reply:
The demo app frequently switches between the UI thread and JS thread. I found these patterns were recurring so having them in the library simplifies the application code.

1. The `object state` parameter for the existing `Send()` and `Post()` callbacks is never used in practice, since it is always more convenient (and doesn't require casting) to capture variables with a lambda.
2. It is often necessary to `await` an operation that runs on the JS thread. The `RunAsync()` methods set up a `TaskCompletionSource` for that.
3. Some overloads allow returning a result to the calling thread, either synchronously from `Run<T>()` or asynchronously from `RunAsync<T>()`.

If you prefer, these methods could be moved to extension methods instead of being directly implemented by `JSSynchronizationContext`.